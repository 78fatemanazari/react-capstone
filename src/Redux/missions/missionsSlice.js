import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const apiUrl = 'https://api.spacexdata.com/v3/missions';

export const fetchMissions = createAsyncThunk('missions/fetchMissions', async () => {
  const response = await axios.get(apiUrl);
  return response.data;
});

const initialState = {
  missions: [],
  loading: false,
  error: null,
  reservedMissions: [],
};

const missionSlice = createSlice({
  name: 'missions',
  initialState,
  reducers: {
    joinMission: (state, { payload }) => {
      state.reservedMissions.push(payload); 
      state.missions = state.missions.map((mission) => {
        if (mission.mission_id === payload) {
          return { ...mission, reserved: true };
        }
        return mission;
      });
    },
    leaveMission: (state, { payload }) => {
      state.reservedMissions = state.reservedMissions.filter((missionId) => missionId !== payload);
      state.missions = state.missions.map((mission) => {
        if (mission.mission_id === payload) {
          return { ...mission, reserved: false };
        }
        return mission;
      });
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchMissions.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchMissions.fulfilled, (state, { payload }) => {
        state.loading = false;
        const result = payload.map((mission) => ({
          mission_id: mission.mission_id,
          mission_name: mission.mission_name,
          description: mission.description,
          reserved: state.reservedMissions.includes(mission.mission_id),
        }));
        state.missions = result;
      })
      .addCase(fetchMissions.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message;
      });
  },
});

export const { joinMission, leaveMission } = missionSlice.actions;
export default missionSlice.reducer;
