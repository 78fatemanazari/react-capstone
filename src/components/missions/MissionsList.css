.missions-container{
    margin: 30px;
}

.table{
  width: 100%;
  border-collapse: collapse;
}

.table th {
  white-space: nowrap;
  text-align: left;
  padding: 10px;
  border: 1px solid #dee2e6;
  border-bottom: 1px;
}

.table td {
  text-align: left;
  padding: 10px;
  border: 1px solid #dee2e6;
  padding-bottom: 20px;
}

.bold-text{
  font-weight: bold;
}

.thin-text{
  font-size: 14px;
}

.status-button{
  border-radius: 5px;
  padding: 3px;
  background-color: gray;
  border: 1px solid gray;
  color: #fff;
  white-space: nowrap;
}

.actions-button{
  border-radius: 3px;
  padding: 7px;
  background-color: #fff;
  white-space: nowrap; 
}

.active-member{
  background-color: dodgerblue;
  border-color: dodgerblue;
  color: #fff;
}

.leave-btn{
  color: red;
  border-color: red;
}

.join-btn{
  padding: 8px;
}