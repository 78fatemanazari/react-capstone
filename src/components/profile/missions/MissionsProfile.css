.no-missions {
  margin-left: 50px;
  font-weight: bold;
  margin-top: 10px;
}

.container {
  display: flex;
}

li {
  list-style: none;
  margin-left: 30px;
}

.missions {
  width: 50%;
}

.joined-mission-list {
  padding: 0;
  width: 90%;
  border-radius: 3px;
  border: 1px solid #dee2e6;
}

.joined-mission-list li {
  padding: 1%;
  margin: 5px;
  font-size: 1.1rem;
  font-weight: 400;
  border-radius: 3px;
  list-style-type: none;
  border: 1px solid #dee2e6;
}
