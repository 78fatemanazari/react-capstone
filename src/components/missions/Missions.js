import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchMissions, joinMission, leaveMission } from '../../Redux/missions/missionsSlice';
import MissionsList from './MissionsList';

const Missions = () => {
  const dispatch = useDispatch();
  const { missions, isLoading, error } = useSelector((state) => state.missions);

  useEffect(() => {
    dispatch(fetchMissions());
  }, [dispatch]);

  const handleJoinMissions = (missionId) => {
    dispatch(joinMission(missionId));
  };

  const handleLeaveMissions = (missionId) => {
    dispatch(leaveMission(missionId));
  };
  return (
    <MissionsList
      missions={missions}
      isLoading={isLoading}
      error={error}
      handleJoinMissions={handleJoinMissions}
      handleLeaveMissions={handleLeaveMissions}
    />
  );
};

export default Missions;
