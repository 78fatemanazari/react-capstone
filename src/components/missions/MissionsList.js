import React from 'react';
import PropTypes from 'prop-types';
import './MissionsList.css';

const MissionsList = ({
  missions, loading, error, handleJoinMissions, handleLeaveMissions,
}) => {
  if (loading) {
    return <div>Loading...</div>;
  }

  if (error) {
    return (
      <div>
        Error:
        {error}
      </div>
    );
  }

  return (
    <div className="missions-container">
      <table className="table">
        <thead>
          <tr>
            <th>Mission</th>
            <th>Description</th>
            <th>Status</th>
            <th>Actions</th>
          </tr>
        </thead>
        <tbody>
          {missions.map((mission, index) => (
            <tr key={mission.mission_id} className={index % 2 === 0 ? 'even' : 'odd'}>
              <td className="bold-text">{mission.mission_name}</td>
              <td className="thin-text">{mission.description}</td>
              <td>
                {mission.reserved ? (
                  <button className="active-member status-button" type="button">
                    Active Member
                  </button>
                ) : (
                  <button className="not-active-member status-button" type="button">
                    NOT A MEMBER
                  </button>
                )}
              </td>
              <td>
                {mission.reserved ? (
                  <button className="actions-button leave-btn" type="button" onClick={() => handleLeaveMissions(mission.mission_id)}>
                    Leave Mission
                  </button>
                ) : (
                  <button 
                    className="actions-button join-btn"
                    type="button"
                    onClick={() => handleJoinMissions(mission.mission_id)}
                  >
                    Join Mission
                  </button>
                )}
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

MissionsList.propTypes = {
  missions: PropTypes.arrayOf(
    PropTypes.shape({
      mission_id: PropTypes.number.isRequired,
      mission_name: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      reserved: PropTypes.bool.isRequired,
    }),
  ).isRequired,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.string,
  handleJoinMissions: PropTypes.func.isRequired,
  handleLeaveMissions: PropTypes.func.isRequired,
};

MissionsList.defaultProps = {
  error: null,
};

export default MissionsList;
