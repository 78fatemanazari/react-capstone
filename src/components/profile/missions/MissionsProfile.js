import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchMissions } from '../../../redux/missions/missionsSlice';
import './MissionsProfile.css';

const MissionsProfile = () => {
  const dispatch = useDispatch();
  const { missions, reservedMissions } = useSelector((state) => state.missions);

  useEffect(() => {
    dispatch(fetchMissions());
  }, [dispatch]);

  const reservedList = missions.filter((mission) => reservedMissions.includes(mission.mission_id));

  return (
    <>
      {reservedList.length === 0 ? (
        <p className="no-missions">You do not have any missions!</p>
      ) : (
        <ul className="missions-list">
          {reservedList.map((mission) => (
            <li key={mission.mission_id}>{mission.mission_name}</li>
          ))}
        </ul>
      )}
    </>
  );
};

export default MissionsProfile;
