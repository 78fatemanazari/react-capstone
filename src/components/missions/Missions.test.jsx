import React from 'react';
import { render, waitFor, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Missions from './Missions';
import { fetchMissions } from '../../Redux/missions/missionsSlice';

const mockStore = configureMockStore([thunk]);

describe('Missions Component', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  test('should display loading state', async () => {
    const store = mockStore({
      missions: {
        missions: [], loading: true, error: null, reservedMissions: [],
      },
    });

    render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );

    store.dispatch(fetchMissions());

    expect(screen.getByText('Loading...').textContent).toBe('Loading...');
  });

  test('should display error state', async () => {
    const store = mockStore({
      missions: {
        missions: [], loading: false, error: true, reservedMissions: [],
      },
    });

    render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );

    store.dispatch(fetchMissions());

    expect(screen.getByText('Error:').textContent).toBe('Error:');
  });

  test('should display content', async () => {
    const store = mockStore({
      missions: {
        missions: [
          {
            mission_id: 1,
            mission_name: 'Mission 1',
            description: 'Mission 1 Description',
            reserved: true,
          },
          {
            mission_id: 2,
            mission_name: 'Mission 2',
            description: 'Mission 2 Description',
            reserved: false,
          },
        ],
        loading: false,
        error: false,
      },
    });

    const { mission } = render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );

    store.dispatch(fetchMissions());

    await waitFor(() => {
      expect(mission).toMatchSnapshot();
    });
  });
});
