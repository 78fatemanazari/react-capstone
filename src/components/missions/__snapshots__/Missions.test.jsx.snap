// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions Component should display content 1`] = `undefined`;
